### PR TITLE
New rule that detects space before parentheses

### PR DESCRIPTION
**Pull request by realmarv**

### CODE REVIEW

File: commitlint.config.ts [hunk 1]
```diff
@@ -207,6 +207,28 @@ module.exports = {
...
+                        if (parenIndex >= 1){
+                            if (headerStr[parenIndex - 1] === ' ') {
+                                offence = true;
+                            }
+                        }    
+                    }
+
+                    return [
+                        !offence,
+                        `Please watch out for leading or ending trailing whitespace`
```
Comment (knocte) [inline commitlint.config.ts:229]: @realmarv well done so far! but we made a mistake here ^ (copy+paste problem hehehe)

File: commitlintplugins.test.ts [hunk 2]
```diff
@@ -205,3 +205,23 @@ test('type-space-after-comma2', () => {
...
+    let typeSpaceBeforeParen2 = runCommitLintOnMsg(commitMsgWithNoSpaceBeforeParen);
+    expect(typeSpaceBeforeParen2.status).toBe(0);
+});
+
+
+test('type-space-before-paren3', () => {
+    let commitMsgWithNoSpaceBeforeParen = "(bar): bla bla bla";
+    let typeSpaceBeforeParen3 = runCommitLintOnMsg(commitMsgWithNoSpaceBeforeParen);
+    expect(typeSpaceBeforeParen3.status).toBe(0);
+});
```
Comment (knocte) [inline commitlintplugins.test.ts:227]: @realmarv nit: also please add EOL before EOF here ^

**Review by knocte — CHANGES_REQUESTED** [review]
See, I added 2 comments to address. After you fix those, I'll merge the PR.

File: commitlint.config.ts [hunk 3]
```diff
@@ -226,7 +226,7 @@ module.exports = {
 
                     return [
                         !offence,
-                        `Please watch out for leading or ending trailing whitespace`
+                        `Please watch out for whitespaces before parentheses`
```
Comment (knocte) [inline commitlint.config.ts:229]: @realmarv mmm, this message might be confusing to the user; think about this case! he writes the commit message `Frontend (SendPage): fix crash (thread issue)`, then she will remove all the spaces before parentheses; but we only care about the first one (because it is inside the scope).

File: commitlint.config.ts [hunk 4]
```diff
@@ -207,6 +207,28 @@ module.exports = {
...
+                        if (parenIndex >= 1){
+                            if (headerStr[parenIndex - 1] === ' ') {
+                                offence = true;
+                            }
+                        }    
+                    }
+
+                    return [
+                        !offence,
+                        `Please watch out for whitespaces before parentheses`
```
Comment (knocte) [inline commitlint.config.ts:229]:
```suggestion
                        `No need to use space before parentheses in the area/scope/sub-area section of the title`
```

**Comment by knocte** [comment]
@realmarv Unfortunately, because I merged the other PR, now this PR has conflicts and cannot be merged, so it needs to be rebased. Please don't use the "git merge" command to do this, you should use "git rebase". If you need help with this, ask in channel please :)

File: commitlint.config.ts (hunk 1 above)
Comment (realmarv) [reply commitlint.config.ts]: Done!

File: commitlintplugins.test.ts (hunk 2 above)
Comment (realmarv) [reply commitlintplugins.test.ts]: Done!

File: commitlint.config.ts (hunk 4 above)
Comment (realmarv) [reply commitlint.config.ts]: Done!

File: commitlint.config.ts (hunk 3 above)
Comment (realmarv) [reply commitlint.config.ts]: Done!

**Comment by realmarv** [comment]
> @realmarv Unfortunately, because I merged the other PR, now this PR has conflicts and cannot be merged, so it needs to be rebased. Please don't use the "git merge" command to do this, you should use "git rebase". If you need help with this, ask in channel please :)

Done!

File: commitlint.config.ts [hunk 5]
```diff
@@ -28,17 +28,17 @@ module.exports = {
...
         'footer-max-line-length': [RuleStatus.Error, 'always', 150],
         'header-max-length': [RuleStatus.Error, 'always', 50],
+        'prefer-slash-over-backslash': [RuleStatus.Error, 'always'],
         'subject-full-stop': [RuleStatus.Error, 'never', '.'],
         'type-empty': [RuleStatus.Warning, 'never'],
         'type-space-after-colon': [RuleStatus.Error, 'always'],
         'subject-lowercase': [RuleStatus.Error, 'always'],
         'type-space-after-comma': [RuleStatus.Error, 'always'],
         'trailing-whitespace': [RuleStatus.Error, 'always'],
-        'prefer-slash-over-backslash': [RuleStatus.Error, 'always'],
```
Comment (knocte) [inline commitlint.config.ts:37]: @realmarv when rebasing, you deleted this by mistake; this line needs to be reinstated before I can merge this PR

Comment (realmarv) [reply commitlint.config.ts:37]: It's in line 31. I thought rules should be ordered alphabetically, but now that I know it's based on their priority, should I put it at the end of the list?

Comment (knocte) [reply commitlint.config.ts:37]: we were writing tests in alphabetical level; then in another PR I pointed out that rules could be in alphabetical level too; but we never talked about configuration of rules being in alphabetical order

Comment (knocte) [reply commitlint.config.ts:37]: typo: level -> order

Comment (realmarv) [reply commitlint.config.ts:37]: Done!

File: commitlintplugins.test.ts [hunk 6]
```diff
@@ -219,3 +219,23 @@ test('type-space-after-comma2', () => {
...
+
+test('type-space-before-paren2', () => {
+    let commitMsgWithNoSpaceBeforeParen = "foo(bar): bla bla bla";
+    let typeSpaceBeforeParen2 = runCommitLintOnMsg(commitMsgWithNoSpaceBeforeParen);
+    expect(typeSpaceBeforeParen2.status).toBe(0);
+});
+
+
+test('type-space-before-paren3', () => {
+    let commitMsgWithNoSpaceBeforeParen = "(bar): bla bla bla";
```
Comment (knocte) [inline commitlintplugins.test.ts:238]: last thing to do in this PR: as we found this case above as an edge case for this rule, because it's an invalid title (as it has sub-area but no area), but shouldn't be covered by this rule, let's add a TODO item in the list that explains we need a rule for this

Comment (realmarv) [reply commitlintplugins.test.ts:238]: Should I remove this test? Because as you said, it's not a correct title, but in this test, we expect the status code to be 0.

Comment (knocte) [reply commitlintplugins.test.ts:238]: good question, maybe we should remove it (or rather rename it) when we implement the new rule that will take care of titles with sub-area but without area

Comment (realmarv) [reply commitlintplugins.test.ts:238]: Right
